torchtune/modules/peft/dora.py

```diff
@@ -104,11 +104,28 @@ def initialize_dora_magnitude(self):
         """
         DoRA initializes the magnitude vector such that its outputs are initially
         identical to standard LoRA's outputs.
+
+        This must be called after loading/initializing base model and LoRA params.
         """
+        if any(
+            [
+                self.weight.is_meta,
+                self.lora_a.weight.is_meta,
+                self.lora_b.weight.is_meta,
+            ]
+        ):
+            raise ValueError(
+                "Cannot initialize DoRA magnitude until after base and LoRA parameters"
+            )
         base_weight = self.weight.to(self.lora_a.weight.dtype)
         lora_weight = self.lora_b.weight @ self.lora_a.weight
-        weight_norm = self._get_weight_norm(base_weight, lora_weight)
-        self.magnitude.copy_(weight_norm)
+
+        magnitude = nn.Parameter(
+            torch.empty_like(self.magnitude, device=self.lora_a.weight.device),
+            requires_grad=self.magnitude.requires_grad,
+        )
+        magnitude = self._get_weight_norm(base_weight, lora_weight)
+        torch.utils.swap_tensors(self.magnitude, magnitude)
```

Comment:
not 100% sure here, but will the original `torch.empty(out_dim)` still be floating around after this? This probably isn't an issue?

Reply:
Yeah I was wondering about this too, need to check myself. Offhand not sure about the implications for refcounting with this utility. But magnitude is just embed_dim number of elements so even if still hanging around it’s pretty small